controllers/tenant/role.go

```diff
@@ -3,73 +3,59 @@ package tenant
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
-func createRole(obj pipeline.Object, data *pipeline.Context) pipeline.Result {
+func reconcileRole(obj pipeline.Object, data *pipeline.Context) pipeline.Result {
 	tenant, ok := obj.(*synv1alpha1.Tenant)
 	if !ok {
 		return pipeline.Result{Err: fmt.Errorf("object is not a tenant")}
 	}
 
-	role, err := newRole(data.Client.Scheme(), tenant)
-	if err != nil {
-		return pipeline.Result{Err: fmt.Errorf("failed to create Role for tenant: %w", err)}
+	rns := []string{tenant.Name}
+	cls := v1alpha1.ClusterList{}
+	if err := data.Client.List(data.Context, &cls,
+		client.InNamespace(tenant.Namespace),
+		client.MatchingFields{"spec.tenantRef.name": tenant.Name},
+	); err != nil {
+		return pipeline.Result{Err: fmt.Errorf("failed to list clusters: %w", err)}
 	}
-
-	err = data.Client.Create(data.Context, role)
-	if err != nil && !errors.IsAlreadyExists(err) {
-		return pipeline.Result{Err: fmt.Errorf("create role: %w", err)}
+	for _, c := range cls.Items {
+		rns = append(rns, c.Name)
 	}
 
-	return pipeline.Result{}
-}
-
-// newRole returns a new Role object.
-// The Role controls access to the tenant and its related clusters.
-func newRole(scheme *runtime.Scheme, tenant *synv1alpha1.Tenant) (*rbacv1.Role, error) {
-	role := &rbacv1.Role{
+	role := rbacv1.Role{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      tenant.Name,
 			Namespace: tenant.Namespace,
 		},
 	}
-	setManagedByLabel(role)
-	if err := controllerutil.SetControllerReference(tenant, role, scheme); err != nil {
-		return nil, err
-	}
-	roleUtil.EnsureRules(role)
-
-	return role, nil
-}
-
-func tenantUpdateRole(obj pipeline.Object, data *pipeline.Context) pipeline.Result {
-	tenant, ok := obj.(*synv1alpha1.Tenant)
-	if !ok {
-		return pipeline.Result{Err: fmt.Errorf("object is not a tenant")}
-	}
-
-	name := types.NamespacedName{Name: tenant.Name, Namespace: tenant.Namespace}
-	role := &rbacv1.Role{}
-	if err := data.Client.Get(data.Context, name, role); err != nil {
-		return pipeline.Result{Err: fmt.Errorf("failed to get role for tenant: %v", err)}
-	}
-
-	if !roleUtil.AddResourceNames(role, tenant.Name) {
-		return pipeline.Result{}
-	}
-
-	if err := data.Client.Update(data.Context, role); err != nil {
-		return pipeline.Result{Err: fmt.Errorf("failed to update role for tenant: %v", err)}
+	op, err := controllerutil.CreateOrUpdate(data.Context, data.Client, &role, func() error {
+		role.Rules = []rbacv1.PolicyRule{
+			{
+				APIGroups:     []string{synv1alpha1.GroupVersion.Group},
+				Verbs:         []string{"get"},
+				Resources:     []string{"tenants", "clusters"},
+				ResourceNames: rns,
+			}, {
+				APIGroups:     []string{synv1alpha1.GroupVersion.Group},
+				Verbs:         []string{"get", "update", "patch"},
+				Resources:     []string{"clusters/status"},
+				ResourceNames: rns,
```

Comment:
Out of curiosity: do I read this right that this allows `get/update/patch` on `clusters/status` for resource name `t-tenant-1234` in addition to the tenant's clusters?

Reply:
Right we can do this nicer now 😅

Reply:
fixed in https://github.com/projectsyn/lieutenant-operator/pull/272/commits/1dba9bf8a82b43dad04f37354b3780c8c9ebcf74.